### PR TITLE
Use filter worker in data-table

### PR DESCRIPTION
**Pull request by bramkragten**
So the UI stays responsive when we are filtering large amounts of data
Before:
before-worker
After:
after-worker

### CODE REVIEW

File: src/components/data-table/ha-data-table.ts [hunk 1]
```diff
@@ -89,50 +99,52 @@ export class HaDataTable extends BaseElement {
   @property({ type: String }) private _filter = "";
   @property({ type: String }) private _sortColumn?: string;
   @property({ type: String }) private _sortDirection: SortingDirection = null;
+  private _filteredData: DataTabelRowData[] = [];
+
+  private _debounceSearch = debounce(
```
Comment (bramkragten) [inline src/components/data-table/ha-data-table.ts]: I'd love to do this without debounce and the 200ms delay, but couldn't get it to work, if you could guide me :-)

File: src/components/data-table/ha-data-table.ts [hunk 2]
```diff
@@ -144,48 +155,22 @@ export class HaDataTable extends BaseElement {
...
-          return sort * -1;
-        }
-        if (valA > valB) {
-          return sort * 1;
-        }
-        return 0;
-      });
+      const column = { ...columns[sortColumn] };
+      delete column.template;
+      return worker.sortData(data, column, direction, sortColumn);
```
Comment (balloob) [inline src/components/data-table/ha-data-table.ts]: Every time we send data to the worker, we need to copy it all over to the worker. Could we do filtering and sorting in one go ? Basically we get a "slim" column definition just for sorting that we can send over.

File: src/components/data-table/ha-data-table.ts [hunk 3]
```diff
@@ -426,7 +411,7 @@ export class HaDataTable extends BaseElement {
   }
 
   private _handleSearchChange(ev: CustomEvent): void {
-    this._filter = ev.detail.value;
+    this._debounceSearch(ev);
```
Comment (balloob) [inline src/components/data-table/ha-data-table.ts]: We can do something like this:

```ts
const curRequest = this.curRequest;
this.curRequest++;

const data = await this.worker.filterAndSortData(this.rawData);
if (this.curRequest != curRequest) return;

// optionally, wait 100ms before starting to process
await new Promise(resolve => setTimeout(resolve, 100))
if (this.curRequest != curRequest) return;

this.data = data;
```

File: src/components/data-table/ha-data-table.ts [hunk 4]
```diff
@@ -89,95 +98,27 @@ export class HaDataTable extends BaseElement {
...
-      });
-    }
-  );
+  protected firstUpdated() {
+    super.firstUpdated();
+    this._worker = sortFilterWorker();
+  }
 
-  protected updated(properties: PropertyValues) {
+  protected async updated(properties: PropertyValues) {
```
Comment (balloob) [inline src/components/data-table/ha-data-table.ts]: `updated` is a lifecycle method that we override, we cannot change the signature to be async.

File: src/components/data-table/ha-data-table.ts (hunk 1 above)
Comment (balloob) [reply src/components/data-table/ha-data-table.ts]: Do we still need this now we also have the other debounce ?

File: src/components/data-table/ha-data-table.ts [hunk 5]
```diff
@@ -192,6 +133,48 @@ export class HaDataTable extends BaseElement {
           break;
         }
       }
+
+      const clonedColumns: DataTabelColumnContainer = deepClone(this.columns);
+      Object.values(clonedColumns).forEach((column: DataTabelColumnData) => {
```
Comment (balloob) [inline src/components/data-table/ha-data-table.ts]: Can we memoize this, so we only create the sorted columns once ?

Comment (bramkragten) [reply src/components/data-table/ha-data-table.ts]: We only do it now if the columns are changed, so memoize would do the same?

**Review by balloob — APPROVED** [review]

File: src/components/data-table/ha-data-table.ts (hunk 5 above)
Comment (balloob) [reply src/components/data-table/ha-data-table.ts]: ok!

**Pull request merged**